src/common/reactNativeProjectHelper.ts

```diff
@@ -10,53 +10,111 @@ import {Package} from "./node/package";
+            );
+        }
+
+        return Q.all(versionPromises).then(packageVersions => ({
+            reactNativeVersion: packageVersions[0],
+            reactNativeWindowsVersion: packageVersions[1] || "",
+        }));
     }
 
-    public static getReactNativeVersionFromProjectPackage(cwd: string): Q.Promise<string> {
-        const rootProjectPackageJson = new Package(cwd);
-        return rootProjectPackageJson.dependencies()
-            .then(dependencies => {
-                if (dependencies["react-native"]) {
-                    return ReactNativeProjectHelper.processVersion(dependencies["react-native"]);
-                }
-                return rootProjectPackageJson.devDependencies()
-                    .then(devDependencies => {
-                        if (devDependencies["react-native"]) {
-                            return ReactNativeProjectHelper.processVersion(devDependencies["react-native"]);
+    public static getReactNativeVersionsFromProjectPackage(cwd: string, isRNWVersion: boolean = false): Q.Promise<RNPackageVersions> {
+        let parsedPackageNames: ParsedPackageName[] = [
+            {
+                packageName: "react-native",
+                useSemverCoerce: true,
+            },
+        ];
+
+        if (isRNWVersion) {
+            parsedPackageNames.push({
+                packageName: "react-native-windows",
+                useSemverCoerce: false,
+            });
+        }
+
+        let versionPromises: Q.Promise<string>[] = [];
+
+        parsedPackageNames.forEach(parsedPackageName => {
+            versionPromises.push((() => {
+                const rootProjectPackageJson = new Package(cwd);
+                return rootProjectPackageJson.dependencies()
+                    .then(dependencies => {
+                        if (dependencies[parsedPackageName.packageName]) {
+                            return ReactNativeProjectHelper.processVersion(dependencies[parsedPackageName.packageName], parsedPackageName.useSemverCoerce);
                         }
+                        return rootProjectPackageJson.devDependencies()
+                            .then(devDependencies => {
+                                if (devDependencies[parsedPackageName.packageName]) {
+                                    return ReactNativeProjectHelper.processVersion(devDependencies[parsedPackageName.packageName], parsedPackageName.useSemverCoerce);
+                                }
+                                return "";
+                            });
+                    })
+                    .catch(err => {
                         return "";
                     });
-            })
-            .catch(err => {
-                return "";
-            });
+                })()
+            );
+        });
+
+        return Q.all(versionPromises).then(packageVersions => ({
+            reactNativeVersion: packageVersions[0],
+            reactNativeWindowsVersion: packageVersions[1] || "",
```

Comment:
As far as I understand, `packageVersions` will always be initialized, since `processVersion` will always return "Invalid Semver" string.

Reply:
Yes, `packageVersions` array  will always exist, but `processVersion` function won't be always called. For example if there isn't required dependency in `package.json`, then an exception will be generated and the promise will return empty string. "Invalid Semver" string will be returned only in case if version of required dependancy doesn't contain any digits  (e.g. `react-native: "test"`).